### ISSUE TITLE
Test and iterate on TS expandable hover apis

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#232685]
To try it:

- Build this TS branch and set it as your workspace TS version: https://github.com/microsoft/TypeScript/pull/59940
- In VS Code `"typescript.experimental.expandableHover": true`

Let's test out TS's current implementation of expandable hovers. The main question is if the UI is good enough as-is, or if it turns out we need something different, which may require API changes

**Comment by aiday-mar** [comment]
Tagging @gabritto about the following

I followed the steps in order to use the TypeScript compiler from the branch. Some feedback I have:

In the vscode repo on line 202 of the file `contentHoverController.ts`, hover on `this._contentWidget`, this  shows an expandable hover. Expand it, this creates a big hover with the expanded type of the content widget. Aside from the fact that the hover is really long, which I am not sure can be avoided, I noticed several things:
  - The hover content is cut off, with ellipsis at the end. Either the editor hover cuts off content or TypeScript truncated the editor text. In this case I would have liked to see the full type.
  - Curiously there is incorrect tokenization/colorization of the code in the hover. I am not sure if this is somehow related to the TreeSitter work that was recently done or if this has always been like this. 
  - As  we previously discussed, it would be good to have clickable links that expand the hover instead of expanding everything at the same time. Expanding everything makes the hover unnecessarily long and hard to navigate. 

https://github.com/user-attachments/assets/40e5a47f-bf12-4f2a-b8bf-ef6f31384b8f

I think it would be good generally to change the API so that we allow linkification of markdown text for the purpose of expansion. Perhaps in the coming iterations, I can develop this, if there is sufficient need for this to be put into the iteration plan.

cc @alexdima

**Comment by aiday-mar** [comment]
I see thanks for letting me know, I'll have a look to remove truncation of hovers

**Comment by aiday-mar** [comment]
Hi I discussed this with my manager. Could we merge and publish the TypeScript branch work to production so we can self-host easily on the work in VS Code?

**Comment by aiday-mar** [comment]
Hi @DanielRosenwasser thanks for letting me know, I will upgrade the TypeScript version

**Comment by aiday-mar** [comment]
Here is the list of feedback I received. I will update this list as more feedback is received:

- [ ] Make the +/- button stick to the bottom of the hover part in such a way that it is always visible. The +/- buttons would scroll with the scrollbar and would clip to the top or the bottom of their respective hover parts.
- [ ] Increase +/- button size

**Comment by mylesmmurphy** [comment]
Hey @aiday-mar and @mjbvz! Really great work on this. I am the owner and maintainer of [Prettify TypeScript](https://github.com/mylesmmurphy/prettify-ts), which provides similar functionality for TS types as an extension today. Needless to say, I am hyped to see this coming built-in as part of VS Code. Really great work 👏 

I've been playing around with TS 5.8 and expanded hover option and have a few questions / feedback 😄 

1. **+/- Location:** The location of the +/- feels a little off from a UX perspective. It feels inconsistent suddenly having a column on the left hand side pushing the information I care about to the right, with a bunch of white space filling up the area my eyes immediately go to. Also, when there's a lot of content that makes a scroll appear, it kind of starts to feel like I'm looking down a bowling alley with the 3 columns, haha. I personally think it would make sense to include these either in the Code Action menu, or on the right hand side, pushed below the scroll bar.

![Image](https://github.com/user-attachments/assets/30dea037-aa05-42d2-800c-56d7b4df35b9)

2. **Intersections:** One of the biggest use cases for my extension is developers like to see intersected types merged into one preview, instead of seeing the `} & {` clutter the view. See the difference between the two options below. 

Current behavior:
![Image](https://github.com/user-attachments/assets/e66cfc2a-a0cd-4563-a545-94c84ce06788)

Desired behavior:
<img width="317" alt="Image" src="https://github.com/user-attachments/assets/dfaa4a77-f6be-4d6c-9d96-6fb5bfd0fd61" />

If this is not desired behavior, I have already created a branch of my existing extension to provide this view instead, which may be the future of the Prettify TS extension. 😄 

3. **Object Prototypes:** A lot of the feedback I receive on my extension is to hide object prototypes, and expand only the information that the end user cares about. In the example below, the highlighted area is the content most users have said they really care about:

![Image](https://github.com/user-attachments/assets/ab14f4c6-999c-4c0b-bd6a-fc27fffce679)

That being said, figuring out and handling all the possible scenarios has been a massive pain in development of Prettify, and I can understand descoping it from this functionality. My extension currently displays the following (depending on user settings):

<img width="283" alt="Image" src="https://github.com/user-attachments/assets/9de86c5c-93f9-4742-9479-afc4d5c360cb" />

4. **Timeline:** If possible, would you be able to share the rough timeline on when this will be released? I am trying to determine if Prettify should continue to be maintained. I see 5.8 is slated for Feb. 25th release, will this functionality be released as experimental as part of that?

Thanks for all your help, and again thank you for your great work on this! This is exactly what many have requested in my project and I am so excited to see it moving in this direction.

**Comment by gabritto** [comment]
@mylesmmurphy thank you for the feedback. The current plan is to have a first implementation of expandable hover ship with TS 5.9, and 5.9 beta should be out mid-April.

**Comment by aiday-mar** [comment]
Hi @mylesmmurphy thank you for the feedback. We will take into account the feedback concerning the positioning of the + and - icons. Once the first implementation is shipped, we will turn it by default on VS Code Insiders.

With the current API, I am not sure if we can have a default verbosity level setting. I will wait for more feedback and perhaps if there is more interest, we could rework the API.

**Comment by insilications** [comment]
I am on 1.98.0-insider with TypeScript 5.9 dev. So right now we can't use the experimental expandable hover?

**Comment by aiday-mar** [comment]
Hi @insilications I believe so yes

**Comment by lindboe** [comment]
Hi, I'm interested in using this feature. Is this possible now that TS 5.9 has been released?

**Comment by gabritto** [comment]
> Hi, I'm interested in using this feature. Is this possible now that TS 5.9 has been released?

@lindboe yes, you should be able to use it already. All you need is to be using TS version 5.9 in vscode on your workspace, and I believe all vscode versions since May already support the feature.

Depending on your vscode version, you may need to check if the `expandable hover` setting is enabled, but it should be enabled by default.

<img width="1180" height="105" alt="Image" src="https://github.com/user-attachments/assets/869c71ca-3c39-425c-888a-ee9f36a4dbd5" />

**Comment by karlhorky** [comment]
Not sure if this is the right venue for UX/UI suggestions, but @runspired had a nice suggestion for the expandable hover over on Bluesky:

> Should just be able to click the symbol/token to expand it or go to it 🫣

https://bsky.app/profile/runspired.com/post/3lvycgp2h3s2r

**Comment by aiday-mar** [comment]
Hi thanks for leaving this comment. From memory, the expandable hovers are available from TypeScript version 5.7 around, so this should be working. Please note that the expandable hovers do not appear on every hover, as the TypeScript server chooses to send them only when it deems it necessary.

**Comment by freeatnet** [comment]
@aiday-mar Are there any other factors affecting when the expandable hovers appear? Attaching a screen recording of a case where they show up when VSCode's TS is selected, but don't show up when the workspace's TS version is enabled, on the same workspace, codebase, and variable, no other changes.

Steps to repro are:
1. Initialize a new Next.js project, `pnpm create next-app@16.0.1 vscode-ts-expandhints-repro-next16` all default config.
2. Pin typescript dependency, `pnpm i typescript@5.9.3` (not strictly required, just making sure it's specific).
3. Open it as a workspace.
4. Go to `app/layout.tsx` and hover over the `metadata` variable. The expandable tooltip appears.
5. Switch to workspace's version of TS.
6. Try step 4 again.

https://github.com/user-attachments/assets/d37032e4-9259-4de4-8f42-0561e98ffc01

**Comment by gabritto** [comment]
> [@aiday-mar](https://github.com/aiday-mar) Are there any other factors affecting when the expandable hovers appear? Attaching a screen recording of a case where they show up when VSCode's TS is selected, but don't show up when the workspace's TS version is enabled, on the same workspace, codebase, and variable, no other changes.
> 
> Steps to repro are:
> 
> 1. Initialize a new Next.js project, `pnpm create next-app@16.0.1 vscode-ts-expandhints-repro-next16` all default config.
> 2. Pin typescript dependency, `pnpm i typescript@5.9.3` (not strictly required, just making sure it's specific).
> 3. Open it as a workspace.
> 4. Go to `app/layout.tsx` and hover over the `metadata` variable. The expandable tooltip appears.
> 5. Switch to workspace's version of TS.
> 6. Try step 4 again.
> 
>  Screen.Recording.2025-11-03.at.15.19.31.mp4

@freeatnet could you open an issue about this on the TypeScript repository and please provide the repro code and TSServer log of the expandable hover attempt? If you don't know how to do that last part, instructions can be found here: https://github.com/microsoft/TypeScript/wiki/Getting-logs-from-TS-Server-in-VS-Code

Also, do you know if you are using any extension that plugs into TSServer? That could be affecting the behavior of hover, though I can't really think of why this would only happen when selecting the workspace version of TS. But you could try disabling all installed extensions and trying this again, just in case.

**Comment by muhamedkarajic** [comment]
I am also interested in this (i think), so I would prefear not to have to click + each time I hover over something if possible by any means!!
